kubernetes-sigs/gcp-compute-persistent-disk-csi-driver
# Use CSI proxy v1 client library

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Adds compatibility with the following versions of the csi-proxy client: disk v1, volume v1, filesystem v1, the plan to upgrade is:

- deploy PD CSI with the csi-proxy v1 (and others) client, keep the beta csi-proxy server on the node, the v1 client would fallback to talking to the beta csi-proxy server
- bump csi-proxy server to v1 on the node, to do this we need to drain the node and restart PD CSI which will then talk to v1 automatically

## Summary of changes:

- add new hostPath mounts to the windows nodes, these mounts are the v1 named pipes, these named pipes don't exist however the implementation is able to understand this fact and it fallbacks to the v1beta2/v1beta1 named pipes
- `safe-mounter_windows.go` was cloned into a v1 and a v1beta version, the differences are the imports (I could also change this to multiple if statements but a clone makes the code easy to follow)
- `NewSafeMounter` creates either a v1 mounter or a v1beta mounter, the `Interface` is casted to the required type on runtime

## Logs of the new implementation:

Test steps:

- create a kubernetes GCE cluster with 2 windows nodes
- choose one of the windows nodes and deploy `csi-proxy.exe`, verify that the v1 named pipes are available

```
PS C:\Users\mauriciopoppe\go\src\github.com\kubernetes-csi\csi-proxy> [System.IO.Directory]::GetFiles("\\.\\pipe\\")
\\.\\pipe\\InitShutdown
\\.\\pipe\\lsass
\\.\\pipe\\ntsvcs
\\.\\pipe\\scerpc
\\.\\pipe\\Winsock2\CatalogChangeListener-364-0
\\.\\pipe\\epmapper
\\.\\pipe\\Winsock2\CatalogChangeListener-1f4-0
\\.\\pipe\\LSM_API_service
\\.\\pipe\\W32TIME_ALT
\\.\\pipe\\eventlog
\\.\\pipe\\Winsock2\CatalogChangeListener-138-0
\\.\\pipe\\wkssvc
\\.\\pipe\\atsvc
\\.\\pipe\\Winsock2\CatalogChangeListener-500-0
\\.\\pipe\\Winsock2\CatalogChangeListener-278-0
\\.\\pipe\\openssh-ssh-agent
\\.\\pipe\\TermSrv_API_service
\\.\\pipe\\Ctx_WinStation_API_service
\\.\\pipe\\SessEnvPublicRpc
\\.\\pipe\\Winsock2\CatalogChangeListener-a7c-0
\\.\\pipe\\srvsvc
\\.\\pipe\\Winsock2\CatalogChangeListener-264-0
\\.\\pipe\\PSHost.132685607009879421.2264.DefaultAppDomain.powershell
\\.\\pipe\\TSVCPIPE-6969a09d-6df7-43e9-8b4d-07ca51b17aeb
\\.\\pipe\\docker_engine
\\.\\pipe\\PSHost.132685607145678539.4928.DefaultAppDomain.powershell
\\.\\pipe\\PIPE_EVENTROOT\CIMV2SCM EVENT PROVIDER
\\.\\pipe\\dockershim
\\.\\pipe\\PSHost.132685608544452389.4896.DefaultAppDomain.powershell
\\.\\pipe\\csi-proxy-filesystem-v1alpha1
\\.\\pipe\\csi-proxy-filesystem-v1beta1
\\.\\pipe\\csi-proxy-filesystem-v1beta2
\\.\\pipe\\csi-proxy-filesystem-v1
\\.\\pipe\\csi-proxy-disk-v1alpha1
\\.\\pipe\\csi-proxy-disk-v1beta1
\\.\\pipe\\csi-proxy-disk-v1beta2
\\.\\pipe\\csi-proxy-disk-v1beta3
\\.\\pipe\\csi-proxy-disk-v1
\\.\\pipe\\csi-proxy-volume-v1alpha1
\\.\\pipe\\csi-proxy-volume-v1beta1
\\.\\pipe\\csi-proxy-volume-v1beta2
\\.\\pipe\\csi-proxy-volume-v1beta3
\\.\\pipe\\csi-proxy-volume-v1
\\.\\pipe\\csi-proxy-smb-v1alpha1
\\.\\pipe\\csi-proxy-smb-v1beta1
\\.\\pipe\\csi-proxy-smb-v1beta2
\\.\\pipe\\csi-proxy-smb-v1
\\.\\pipe\\csi-proxy-system-v1alpha1
\\.\\pipe\\csi-proxy-iscsi-v1alpha1
\\.\\pipe\\csi-proxy-iscsi-v1alpha2
```

- compile PD CSI driver and deploy it the cluster, check the logs of PD CSI driver in both windows nodes

```
# windows node without v1 pipes enabled
I0619 08:14:54.770087    1504 main.go:73] Driver vendor version latest
I0619 08:14:54.843225    1504 safe-mounter_windows.go:35] failed to connect to csi-proxy v1 with error=open \\.\\pipe\\csi-proxy-filesystem-v1: The system cannot find the file specified., will try w
I0619 08:14:54.843225    1504 safe-mounter_windows.go:39] using CSIProxyMounterV1Beta, API Versions Disk: v1beta2, Filesystem: v1beta1, Volume: v1beta1

# windows node with v1 pipes enabled
I0619 08:14:54.500483    3260 main.go:73] Driver vendor version latest
I0619 08:14:54.503411    3260 safe-mounter_windows.go:28] using CSIProxyMounterV1, API Versions Disk: v1, Filesystem: v1, Volume: v1
```

- Test on v1beta: Untaint the v1beta node and taint the v1 node and run the `External Storage` suite from k/k (compatibility with the existing csi-server)

```
GINKGO_PARALLEL_NODES=4 ./hack/ginkgo-e2e.sh --ginkgo.focus="External.*Storage" --ginkgo.skip="\[Disruptive\]|\[Serial\]|\[LinuxOnly\]" --allowed-not-ready-nodes=10 --storage.testdriver="/usr/local/google/home/mauriciopoppe/go/src/sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/test/k8s-integration/config/test-config-windows.yaml" --node-os-distro=windows

Ran 36 of 6637 Specs in 2833.752 seconds
SUCCESS! -- 36 Passed | 0 Failed | 0 Pending | 6601 Skipped
```

- Test on v1: Untaint the v1 node and taint v1beta node and run the `External Storage` suite from k/k

```
GINKGO_PARALLEL_NODES=4 ./hack/ginkgo-e2e.sh --ginkgo.focus="External.*Storage" --ginkgo.skip="\[Disruptive\]|\[Serial\]|\[LinuxOnly\]" --allowed-not-ready-nodes=10 --storage.testdriver="/usr/local/google/home/mauriciopoppe/go/src/sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/test/k8s-integration/config/test-config-windows.yaml" --node-os-distro=windows

Ran 36 of 6637 Specs in 2970.219 seconds
SUCCESS! -- 36 Passed | 0 Failed | 0 Pending | 6601 Skipped
```

- (pending) Run the optional prow job e2e `External Storage` testsuite setting csi-proxy v1 in k/k https://github.com/kubernetes/test-infra/pull/22636

### Next changes:

- bump csi-proxy from v1-rc to v1

**Does this PR introduce a user-facing change?**:

```release-note
Bumped csi-proxy client library to v1.0.0
```

/assign @jingxu97 @msau42
